docs/design/2021-02-09-ttl-table.md

```diff
@@ -0,0 +1,45 @@
+
+## Proposal
+Introduce new `TTL` and `TTL_GRANULARITY` table options when creating or altering tables. By specifying `TTL`, users can expect expired data to be removed automatically. Additionally, users can also choose from `ROW` or `PARTITION` for `TTL_GRANULARITY` option to trade off collect granularity and cost to run such garbage collection. `ROW` mode evaluates expiry time against each row of data and reclaim space at row basis. This gives us the finest granularity and allows the most accurate expiration timing. `PARTITION` mode on the other hand partitions data according to its last update time. A background timer is going to rollover partitions and truncate all expired data within the oldest partition all at once.
+
+## Rationale
+Users can choose to store expiry time explicitly in the table schema and run delete on such tables periodically to reclaim space. This approach not only post burden on users but also introduced unnecessary load on TiDB to execute such updates and run GC at background to eventually reclaim space. The extra load is especially high when there is lots of data expiring within a short period of time.
+Another way to quickly remove expired data is managing all data in a carefully designed partition table. Bucketing data according to their update time makes it possible to quickly delete all expired records within a partition with a simple `TRUNCATE`. Compared to TTL Table's approach, these two options are error prone and suboptimal.
+
+## Compatibility and Migration Plan
+Due to the fact that TiDB by itself prohibits either converting ordinary table to partition table or between different partition table types, it is only possible to convert existing table to TTL table with `ROW` granularity. A TTL table with `PARTITION` granularity is implemented as a special partition table therefore conflicts with any other partition table type. Newly created tables can choose either `ROW` or `PARTITION` as a trade off between accuracy of reclaiming time and efficiency of garbage collecting expired data.
+
+## Implementation
+
+### `ROW`
+By associating TTL configuration with the key range of a TTL table in PD and distributing such configuration to all TiKV instances within the cluster, TiKV can utilize TTL settings during GC process to collect data that are expired. To avoid inconsistency caused by reclaiming progress difference between index and actual record data. The lifetime of record range is one GC interval longer than it is for index range, therefore TiDB will not see missing record data during table lookup.
```

Comment:
could you elaborate more about:

1. How to store TTL definitions in the table schema, what needs to be changed.
2. How does TiDB communicate with PD, the API/protobuf design
3. How does PD communicate with TiKV, the API/protobuf design
4. How to ensure stale rows are purged in TiKV, the internal design of TiKV, or the API to use in TiKV.

Reply:
Sure, Iet me add more about these.

Reply:
1. Changes to model.TableInfo
```go
@@ -298,6 +298,11 @@ type TableInfo struct {

        // TiFlashReplica means the TiFlash replica info.
        TiFlashReplica *TiFlashReplicaInfo `json:"tiflash_replica"`
+
+       // TTL
+       TTL                 time.Duration
+       TTLByRow            bool
+       NextTTLTruncateTime time.Time
 }
```

Reply:
Changes to pdpb.proto to specify TTL configuration for ranges
```protobuf
message RangeTTL {
    bytes start_key = 1;
    bytes end_key = 2;
    uint64 TTL = 3;
    bytes user_data = 4;
    bool add_gc_interval = 5; // delay TTL by adding GC interval
}
service PD {
    rpc AddRangeTTL(AddRangeTTLRequest) returns (AddRangeTTLResponse) {}
    rpc DeleteRangeTTL(DeleteRangeTTLRequest) returns (DeleteRangeTTLResponse) {}
    rpc GetRangeTTL(GetRangeTTLRequest) returns (GetRangeTTLResponse) {}
    rpc GetAllRangeTTL(GetAllRangeTTLRequest) returns (GetAllRangeTTLResponse) {}
}
message AddRangeTTLRequest {
    RequestHeader header = 1;
    repeated RangeTTL TTL = 2;
}
message AddRangeTTLResponse {
    ResponseHeader header = 1;
}
message DeleteRangeTTLRequest {
    RequestHeader header = 1;
    bytes start_key = 2;
    bytes end_key = 3;
}
message DeleteRangeTTLResponse {
    ResponseHeader header = 1;
}
message GetRangeTTLRequest {
    RequestHeader header = 1;
    bytes start_key = 2;
    bytes end_key = 3;
}
message GetRangeTTLResponse {
    ResponseHeader header = 1;
    RangeTTL TTL = 2;
}
message GetAllRangeTTLRequest {
    RequestHeader header = 1;
}
message GetAllRangeTTLResponse {
    ResponseHeader header = 1;
    repeated RangeTTL TTL = 2;
}
message GetGCSafePointResponse {
    ResponseHeader header = 1;
    uint64 safe_point = 2;
    repeated RangeTTL range_TTL = 3;
    uint64 now = 4;
}
```
This is what we have in hackathon to quickly make it work. We should definitely make it better.

Reply:
There are two new fields in GetGCSafePointResponse which TiKV uses to get safe_point for every run of GC. These two new fields can tell TiKV what is the current time on PD and the ranges that have TTL enabled. During GC process, if a KV pair is older than safe_point and is expired, it is collected no matter not matter what to reclaim the space.

Reply:
There are two new fields in GetGCSafePointResponse which TiKV uses to get safe_point for every run of GC. These two new fields can tell TiKV what is the current time on PD and the ranges that have TTL enabled. During GC process, if a KV pair is older than safe_point and is expired, it is collected no matter not matter what to reclaim the space.